# Info Challenge 2024 - Sustainability Datset
By Ryan Kim, Ryan Deppe & Maria Lanasa
\
2/24/2024

## Introduction
The challenge being presented is dealing with GreenTerp's data set dealing with

* academic year info is from AY2018-2019 to AY2022-2023
* student can submit anually, students can register once and be done with it, but to be certified you have to verify certification annually
* There are edge cases where people have certified without registering, and those are the red flags we are looking for
* registration and certification requirements change over time
* all data is self reported

## Question:

Our team is concerned with the trends following the data of GreenTerp's submission forms. In detail, we are looking for trends in data on a yearly basis to see if there any concerning processes and analyzing how to improve said processes. Additionally, based on said trends are we able to recommend GreenTerp with what to do for the present and future?

\
Thus we are concerned with the following:

*   How likely are people who register to certify (the conversion rate)? What are the yearly conversion rates?
*   How many members are 'red flags'? On average, how many red flags are seen on yearly basis?




## Dictionary & Brief Explanation of Process
'Red flag' - people who submit certification forms but have never registered
\
Registration - registration forms to join GreenTerp
\
Certification - certificating to receive a reward


GreenTerp processes forms through a Google spreadsheet where each form is either a registration form or a certification form. Registration forms allow a person to join the GreenTerp's club while the certification form allows a person to be elgible for a prize granted that they follow the guidelines for certification. Typically, a person is supposed

## Methods
Firstly we look to import necessary libraries to utilize the data in an efficent manner, our team utilzied Python and R to parse through data, and analyze said data into presentable formats.

In [ ]:
import numpy as np
from collections import Counter
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


### What is the conversion rate from registration to certification?

To answer this question, the equation we are using is

```
conversion_rate = (# of certification forms)/(# of registered forms)
```
Keep in mind that this does not exclude 'red flags' as stated previously; people who have certified first and never registered to begin with.

In [ ]:
# Loading an array of all form types
allForms = np.loadtxt('/content/drive/MyDrive/Info Challenge 2024/Data_Level3_GreenTerp - Cleaned.csv', skiprows=1, usecols=2, delimiter=',',dtype=object)
certified = []
registered = []

#Finding all of the locations in the spreadsheet where it is a registration or a certification form
for x in range(0,len(allForms)):
  if allForms[x] == 'Certification':
    certified.append(x)
  elif allForms[x] == 'Registration':
    registered.append(x)

print(f'Conversion rate of registered to certified is {len(certified)/len(registered)*100}%')


Conversion rate of registered to certified is 46.502331778814124%


But is there human error at play? Could someone register/certify twice in a year causing the numbers to look much more worse? Additionally, people need to certify year after year, so we need to find the number of unique certifications versus unique registrations.

In [ ]:
#Loading student's IDs
ids = np.loadtxt('/content/drive/MyDrive/Info Challenge 2024/Data_Level3_GreenTerp - Cleaned.csv', skiprows=1, usecols=9, delimiter=',',dtype=object)

# Creating a dictionary of ID to form type Key:value pairs, ID: [# of registrations, # of certifications]
id_form = dict()

# A function that changes a dictionary's value pairs based on the key
def changeDict(loc, dic):
  if loc in certified:
    dic[ids[loc]][1] +=1
  else:
    dic[ids[loc]][0] +=1
  return dic

# Now we parse through the dataset to assign each ID a conversion rate
for x in range(0,len(ids)):
  if ids[x] in id_form.keys():
    id_form = changeDict(x,id_form)
  else:
    id_form[ids[x]] = [0,0]
    id_form = changeDict(x,id_form)


# Now finding the number of unique registrations and unique certifications
def findRegAndCerts(dSet):
  num_reg = 0
  num_cert = 0
  for id in dSet.keys():
    if id_form[id][0] >= 1:
      num_reg+=1

    if dSet[id][1] >= 1:
      num_cert +=1
  return [num_reg,num_cert]

regOrCert = findRegAndCerts(id_form)
print(f'The unique conversion rate from registering to certifying is {regOrCert[1]/regOrCert[0]*100}%')

The unique conversion rate from registering to certifying is 48.495200911013505%


We can see that the conversion rate has increased than before, showcasing the impact of human error on given data. However we have to conversion
### What are the yearly conversion rates?
Now we do the same as before, but now seperate it based on what year the forms were submitted. \
As stated before we are concerned with the following years:


*   2018-2019
*   2019-2020
* 2021-2022
* 2022-2023



In [ ]:
# Intializing variables, again it's a dictionary with a key:value pair of ID : list[# of registrations, # of certifications]
y2018 = dict()
y2019 = dict()
y2021 = dict()
y2022 = dict()

# Finding year indexes
years = np.loadtxt('/content/drive/MyDrive/Info Challenge 2024/Data_Level3_GreenTerp - Cleaned.csv', skiprows=1, usecols=0, delimiter=',',dtype=object)
yearIndex = [0]
for x in range(1,len(years)-1):
  if years[x] != years[x+1]:
    yearIndex.append(x+1)

yearIndex.append(len(years)-1)

# Function to go through each year
def yearAdd(yearDict, index):

  for x in range (yearIndex[index],yearIndex[index+1]+1):
    if ids[x] in yearDict.keys():
      yearDict = changeDict(x,yearDict)
    else:
      yearDict[ids[x]] = [0,0]
      yearDict = changeDict(x,yearDict)

  return yearDict

# Adding all of the values to the dictionary
y2018 = yearAdd(y2018,3)
y2019 = yearAdd(y2019,2)
y2021 = yearAdd(y2021,1)
y2022 = yearAdd(y2022, 0)

Now we can try looking for the yearly conversion rates.

In [ ]:
y18 = findRegAndCerts(y2018)
y19 = findRegAndCerts(y2019)
y21 = findRegAndCerts(y2021)
y22 = findRegAndCerts(y2022)

#Printing all conversion rates
print(f'The conversion rate from registering to certifying in 2018-2019 was {y18[1]/y18[0]*100}% with {y18[1]} certifications and {y18[0]} registrations')
print(f'The conversion rate from registering to certifying in 2019-2020 was {y19[1]/y19[0]*100}% with {y19[1]} certifications and {y19[0]} registrations')
print(f'The conversion rate from registering to certifying in 2021-2022 was {y21[1]/y21[0]*100}% with {y21[1]} certifications and {y21[0]} registrations')
print(f'The conversion rate from registering to certifying in 2022-2023 was {y22[1]/y22[0]*100}% with {y22[1]} certifications and {y22[0]} registrations')

The conversion rate from registering to certifying in 2018-2019 was 49.60691823899371% with 1262 certifications and 2544 registrations
The conversion rate from registering to certifying in 2019-2020 was 44.52415112386418% with 931 certifications and 2091 registrations
The conversion rate from registering to certifying in 2021-2022 was 50.93765838824126% with 1005 certifications and 1973 registrations
The conversion rate from registering to certifying in 2022-2023 was 19.50718685831622% with 95 certifications and 487 registrations


But what about the conversion rate for new members who register and certify within the year? \
What about accounting for 'red flags' that don't register and only certified?

In [ ]:
#Finding all instances of IDs that register and certify within the year
#Also removes 'red flags'
def findUnique(idDict):
  numReg = 0
  numCert = 0
  for id in idDict:
    if idDict[id][0] >= 1:
      numReg += 1
      if idDict[id][1] >= 1:
        numCert +=1
  return [numReg,numCert]

y18 = findUnique(y2018)
y19 = findUnique(y2019)
y21 = findUnique(y2021)
y22 = findUnique(y2022)

print(f'The conversion rate from registering to certifying within 2018-2019 was {y18[1]/y18[0]*100}% with {y18[1]} certifications and {y18[0]} registrations')
print(f'The conversion rate from registering to certifying within 2019-2020 was {y19[1]/y19[0]*100}% with {y19[1]} certifications and {y19[0]} registrations')
print(f'The conversion rate from registering to certifying within 2021-2022 was {y21[1]/y21[0]*100}% with {y21[1]} certifications and {y21[0]} registrations')
print(f'The conversion rate from registering to certifying within 2022-2023 was {y22[1]/y22[0]*100}% with {y22[1]} certifications and {y22[0]} registrations')

The conversion rate from registering to certifying within 2018-2019 was 41.914722445695894% with 1042 certifications and 2486 registrations
The conversion rate from registering to certifying within 2019-2020 was 44.52415112386418% with 931 certifications and 2091 registrations
The conversion rate from registering to certifying within 2021-2022 was 38.883034773445736% with 738 certifications and 1898 registrations
The conversion rate from registering to certifying within 2022-2023 was 9.782608695652174% with 45 certifications and 460 registrations


With the conversion rate having an average of `48.495200911013505%` over the entire dataset, the fluctuations in the conversion rates for the years up to 2021 were around the average. However we see a steep decline in year 2023, with a outlier of a conversion rate of `19.50718685831622%`. Keep in mind that the team was told that the dataset was from November of 2023, meaning that the 2022-2023 season had fully concluded by that point.

Additionally from the results, we see that while in past years GreenTerp's registrations have been around the 2000 member mark and about 1000 certifications, in the year 2022-2023 we see only 487 registrations and 95 certifications.


Even after looking at only new member data and accounting for 'red flags', we still find that there has been a concerning decrease in both registered and certified members in 2023-2023.




### How many 'red flags' are really there?
'Red flags' are people who have only certified and never have registered over a time span from 2019-2023, to find this type of person within the data set is fairly simple: find a list of people who have certified and compare it to those who have registered, if a person who is certified is not registered, they are a 'red flag'.

In [ ]:
#Checking dictionary to find people who are certified but not registered
redflags = list()
for id in id_form:
  if id_form[id][1] >= 1 and id_form[id][0] == 0:
    redflags.append(id)

print(f'Number of red flags: {len(redflags)}\nNumber of unique members: {len(id_form)}')
print(f'Percentage of all unique members who are considered as red flags {len(redflags)/len(id_form)*100}%')

Number of red flags: 375
Number of unique members: 6522
Percentage of all unique members who are considered as red flags 5.749770009199632%
